File: js/src/list/linkedList.js
function LinkedList(value) {
    this.next = null;
    this.value = value !== undefined ? value : null; // set the value to null if none is passed in
}

LinkedList.prototype = (function() {
    function add(value, index) {
        if(this.value === null) {
            this.value = value;
        } else if(index !== 0 && !index) {
            if(this.next) {
                this.next.add(value);
            } else {
                this.next = new LinkedList(value);
            }
        } else {
            if(invalidIndex(index, this.length())) {
                console.log('warning: invalid index ' + index + ' passed to LinkedListNode.add()');
                console.log(length());
                return;
            }

            if(index === 0) {
                if(this.next != null) {
                    var tmp = this.next;
                    this.next = new LinkedList(value);
                    this.next.next = tmp;
                } else {
                    this.next = new LinkedList(value);
                }
            } else {
                this.next.add(value, index - 1);
            }
        }
    }

    function get(index) {
        if(invalidIndex(index, this.length())) {
            console.log('warning: invalid index ' + index + ' passed to LinkedListNode.get()');
            return null;
        } else if(index === 0) {
            return this.value;
        } else {
            return this.next.get(index - 1);
        }
    }

    function set(index, value) {
        if(invalidIndex(index, this.length())) {
            console.log('warning: invalid index ' + index + ' passed to LinkedListNode.set()');
        } else if(index === 0) {
            this.value = value;
        } else {
            this.next.set(index - 1, value);
        }
    }

    function remove(index) {
        if(invalidIndex(index, this.length())) {
            console.log('warning: invalid index ' + index + ' passed to LinkedListNode.remove()');
        } else if(index === 0) {
            this.value = this.next.value;
            this.next = this.next.next;
        } else {
            if(index === 1) {
                this.next = this.next.next;
            } else {
                this.next.remove(index - 1);
            }
        }
    }

    function length() {
        // null value is how we show an empty list
        if(this.value === null) {
            return 0;
        } else if(this.next) {
            return this.next.length() + 1;
        } else {
            return 1;
        }
    }

    function contains(target, comparator) {
        if(this.equals(target, comparator)) {
            return true;
        } else if (this.next) {
            return this.next.contains(target, comparator);
        } else {
            return false;
        }
    }

    function indexOf(target, comparator) {
        if(this.equals(target, comparator)) {
            return 0;
        } else if(this.next) {
            var index = this.next.indexOf(target, comparator);
            return index === -1 ? index : index + 1; // if it isn't found, just return -1
        } else {
            return -1;
        }
    }

    function equals(value, comparator) {
        return (comparator && comparator(this.value, value)) || (!comparator && this.value === value);
    }

    function invalidIndex(index, len) {
        return typeof(index) !== 'number' || index < 0 || index >= len;
    }

    return {
        add: add,
        get: get,
        set: set,
        remove: remove,
        length: length,
        contains: contains,
        indexOf: indexOf,
        equals: equals
    }
}.call(this))

File: Java/src/list/LinkedList.java
package list;

/**
 * Generic singly linked list implementation
 *
 * Uses a flag to handle the list being empty, since it's
 * safer than using a null check on value
 *
 * This implementation uses recursion
 * instead of iterative approaches to various methods
 */
public class LinkedList<T> {
    private T value;
    private LinkedList<T> next;
    private boolean empty;

    public LinkedList() { empty = true; }

    private LinkedList(T value) {
        this.empty = false;
        this.value = value;
    }

    public int length() {
        if(empty) {
            return 0;
        } else if(next == null) {
            return 1;
        } else {
            return 1 + next.length();
        }
    }

    public boolean isEmpty() {
        return empty;
    }

    /**
     * get the value at the given index
     */
    public T get(int index) {
        if(index < 0 || index > this.length() - 1) {
            throw new IndexOutOfBoundsException();
        } else if(index == 0) {
            return this.value;
        } else {
            return next.get(index - 1);
        }
    }

    /**
     * append a value to the list
     */
    public void add(T value) {
        if(empty) {
            this.value = value;
            empty = false;
        } else if(next == null) {
            next = new LinkedList<>(value);
        } else {
            next.add(value);
        }
    }

    /**
     * insert a value before the given index
     */
    public void add(T value, int index) {
        if(index < 0 || index > this.length()) {
            throw new IndexOutOfBoundsException();
        } else if(index == 0){
            this.value = value;
            empty = false; // in case this was the first insert
        } else if(index == 1) {
            LinkedList<T> tmp = next;
            next = new LinkedList<>(value);
            next.next = tmp;
        } else {
            next.add(value, index - 1);
        }
    }

    /**
     * set the value at the given index
     */
    public void set(T value, int index) {
        if(index < 0 || index > this.length() - 1) {
            throw new IndexOutOfBoundsException();
        } else if(index == 0) {
            this.value = value;
        } else {
            next.set(value, index - 1);
        }
    }

    /**
     * get the index of the given value, -1 if not found
     */
    public int indexOf(T target) {
        if(!this.contains(target)) {
            return -1;
        } else if(this.value.equals(target)) {
            return 0;
        } else {
            return next.indexOf(target) + 1;
        }
    }

    /**
     * check if the given value is contained in the list
     */
    public boolean contains(T target) {
        return this.value.equals(target) || (this.next != null && this.next.contains(target));
    }
}
